### ISSUE TITLE
SslStream.WriteAsync hangs after updating to CoreFx 4.4

### ISSUE THREAD

**Issue by cesarbs** [dotnet/corefx#14698]
We've recently updated package references in Kestrel to CoreFx 4.4 https://github.com/aspnet/KestrelHttpServer/commit/5b8f7c2b2be5744edac8ce3c5a2dea86ed65b8ee.

After this update, Kestrel can't serve HTTPS anymore. We have a stream that wraps an `SslStream`, and it's hanging when writing here:

https://github.com/aspnet/KestrelHttpServer/blob/5b8f7c2b2be5744edac8ce3c5a2dea86ed65b8ee/src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/StreamSocketOutput.cs#L59

Reproing is straightforward: just run the Kestrel sample app (under `samples/SampleApp` in the repo) and hit it with any client, and you will see the hang.

This only happens on .NET Core. Desktop .NET works fine.

**Comment by benaadams** [comment]
Effecting corefx change is likely to be "SslStream override FlushAsync" https://github.com/dotnet/corefx/pull/14035 

Does it still happen after this Kestrel PR "Implement Stream Flush+FlushAsync fully"? https://github.com/aspnet/KestrelHttpServer/pull/1271?

**Comment by cesarbs** [comment]
@benaadams Yes.

**Comment by benaadams** [comment]
Unrelated to that; lot of dipping through Stream; seems to be blocked on a sync SemaphoneSlim Wait

![](https://aoa.blob.core.windows.net/aspnet/sslstream-block.png)

**Comment by benaadams** [comment]
Feels like a regression of https://github.com/dotnet/corefx/pull/7819; new apis coming back changing the overload resolutions? Are Stream Begin/End back in coreclr?

/cc @CIPop @stephentoub

**Comment by karelz** [comment]
cc: @Priya91

**Comment by CIPop** [comment]
> Are Stream Begin/End back in coreclr?

@JeremyKuhne, @ianhays do you know if these got added back? (in corefx [System.IO typeforwards Stream](https://github.com/dotnet/corefx/blob/master/src/System.IO/ref/System.IO.Manual.cs))

**Comment by ianhays** [comment]
>@JeremyKuhne, @ianhays do you know if these got added back? (in corefx System.IO typeforwards Stream)

Yes, Begin/End are back in Stream.

**Comment by karelz** [comment]
How did the overload resolutions worked on Desktop? Or is it unique .NET Core-only problem?

**Comment by danmosemsft** [comment]
Seems blocking Kestrel.

**Comment by CIPop** [comment]
> How did the overload resolutions worked on Desktop? Or is it unique .NET Core-only problem?

It's unique to .Net Core where the APM Apis for SslStream have been implemented using extension methods (since the base versions didn't exist).

**Comment by cesarbs** [comment]
Is there an ETA for this? We had to disable a number of tests to get our CI infrastructure moving, and HTTPS is completely broken on Kestrel.

**Comment by CIPop** [comment]
@cesarbs I don't think we've root-caused the issue although @benaadams hints it might be the APM extension methods. 

Lots of folks are on vacation these days and I'm already fully booked on other high priority items.
If you have some time, we would appreciate some help in investigating and possibly fixing (even if temporarily) the issue to unblock your team.

**Comment by cesarbs** [comment]
Haven't exactly found the root cause, but at least I found a workaround for Kestrel. If I put a "pass through" stream before `SslStream` that does _not_ implement `BeginWrite`/`EndWrite`, I avoid [this](https://github.com/dotnet/coreclr/blob/1cfbb8d79d65ea972eed8d73458fdf93697d031e/src/mscorlib/src/System/IO/Stream.cs#L426) code path in `Stream`, and HTTPS works again.

**Comment by cesarbs** [comment]
I've found a workaround for Kestrel, but I don't know what exactly is going on. I can get HTTPS working again by placing the following stream before `SslStream`:

```csharp
    public class PassThroughStream : Stream
    {
        private readonly Stream _inner;

        public PassThroughStream(Stream inner)
        {
            _inner = inner;
        }

        public override bool CanRead => _inner.CanRead;

        public override bool CanSeek => _inner.CanSeek;

        public override bool CanWrite => _inner.CanWrite;

        public override long Length => _inner.Length;

        public override long Position
        {
            get
            {
                return _inner.Position;
            }

            set
            {
                _inner.Position = value;
            }
        }

        public override void Flush() => _inner.Flush();

        public override int Read(byte[] buffer, int offset, int count) => _inner.Read(buffer, offset, count);

        public override long Seek(long offset, SeekOrigin origin) => _inner.Seek(offset, origin);

        public override void SetLength(long value) => _inner.SetLength(value);

        public override void Write(byte[] buffer, int offset, int count) => _inner.Write(buffer, offset, count);

        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
        {
            return _inner.WriteAsync(buffer, offset, count, cancellationToken);
        }
    }
```

Note the override to `WriteAsync` - without it, it still hangs. Now what's interesting is that if I override `ReadAsync` merely as

```csharp
    public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
    {
        return _inner.ReadAsync(buffer, offset, count, cancellationToken);
    }
```

The issue repros again.

Also, I get rid of the hang if I place that stream anywhere along a chain of streams coming before `SslStream`, as long as it does not override `ReadAsync`.

**Comment by karelz** [comment]
@cesarbs @benaadams can you please help me understand the root cause? (if you know it)
It looks like some added overloads lead to 'wrong' resolutions on Kestrel side -- naively I would guess Kestrel can use explicitly the right overaloads, can't it? (I may not understand the problem enough, so please bear with me if that's the case)

@cesarbs can you please help me understand how much is this blocking Kestrel? Is it blocking daily builds or existing RTM/RC releases? When was the regression introduced?
Did you consider options to not consume latest CoreFX? Is that an option at all?

**Comment by benaadams** [comment]
.NET Full Framework 1.1+ had async methods `Stream.BeginWrite`/`Stream.EndWrite` and `Stream.BeginRead`/`Stream.EndRead` on the base class. This approach to async was depreciated when `Task`s became a thing and the new Task based methods `Stream.WriteAsync` and `Stream.ReadAsync` became the preferred method of doing async.

The base classes implementation of `BeginX`/`EndX` are sync methods to support Streams that don't actually have async methods. A lot of the conversions for the historic Streams newer `XXXAsync` methods have wrapped these `BeginX`/`EndX` methods to achieve the newer async paths; using helpers that do the conversion.

As these methods were deprecated they were not brought to .NET Core. To reuse the same code that expected `BeginX`/`EndX` without doing lots of rewriting to directly implement `WriteAsync`/`ReadAsync`; which would be additional replumbing and testing to end up with the same result (although it would be cleaner); extension methods were created to behave as the `BeginX`/`EndX` methods. So all was fine...

However, as part of returning the apis for .NET Standard 2.0; this has meant the base class implementations for `BeginX`/`EndX` have reappeared. The virtual base class methods take precedence over extension methods; so the behaviour has moved back to sync-blocking on the async methods; rather than actually being async.

Hence this issue has arisen... 😢 

The fastest way to fix it is likely to copy paste the extension methods code into SslStream/NetworkStream etc. The best way to fix it would be to actually implement the `WriteAsync`/`ReadAsync` methods - though likely that would also be more risky.

**Comment by karelz** [comment]
Thanks @benaadams! That starts to make more sense to me. (disclaimer: I didn't verify your technical details, but your write up makes sense :))
In general -- I am interested more in the "right fix" from compat perspective (compat with Desktop/Xamarin vs. previous .NET Core versions). And I am also interested in: What is the right thing for the platform?
I would jump to fastest solutions only if we absolutely necessarily have to -- which I hope should not be the case in daily builds. If daily Kestrel builds are truly blocked without simple workaround (I still hope it's not the case), then we could just back out the change which broke it. However, the change is in CoreFX for a few weeks at minimum, which makes me question how quickly proper fix will be validated in Kestrel or how quickly can the unblocking fix be consumed ... @cesarbs can you provide more background info here?

@Priya91 can you please look into this one?

**Comment by cesarbs** [comment]
@karelz This is blocking Kestrel in that HTTPS is not working at all - whenever it tries to write a response (which eventually ends up in `SslStream.WriteAsync`), it just hangs forever. So we had to disable all of our tests that use HTTPS to at least get the CI running.

I've been debugging this for the past couple of hours. The problem seems to arise whenever both `ReadAsync` and `WriteAsync` are called on a stream that overrides `BeginRead`/`EndRead` and `BeginWrite`/`EndWrite`, which is the case of `SslStream`. If I put my workaround above in place, `ReadAsync` is called on `PassThroughStream` which takes this [optimized path](https://github.com/dotnet/coreclr/blob/1cfbb8d79d65ea972eed8d73458fdf93697d031e/src/mscorlib/src/System/IO/Stream.cs#L426), while `WriteAsync` uses `SslStream`'s `WriteAsync` which will hit [the other path](https://github.com/dotnet/coreclr/blob/1cfbb8d79d65ea972eed8d73458fdf93697d031e/src/mscorlib/src/System/IO/Stream.cs#L430). I've also verified that the fix works by reversing this i.e. having `PassThroughStream` call the inner stream's `ReadAsync` but the base `WriteAsync`.

I haven't gotten to the bottom of this though.

I'll try to write a smaller repro apart from Kestrel. Hopefully that will make it less noisy to debug.

**Comment by karelz** [comment]
@cesarbs it is not clear to me, why Kestrel is blocked now. The breaking APIs are in CoreFX for weeks at least. Do you have an explanation here?

**Comment by cesarbs** [comment]
We've taken new CoreFX packages recently, that's when it broke. cc @pranavkm for info on when was the previous time we had taken them.

**Comment by pranavkm** [comment]
We recently got around to picking up the 4.4.0 CoreFx packages. We were blocked due to some CLI issues that finally got resolved this past week.

**Comment by karelz** [comment]
I see. One temporary solution would be to back out that change, if Kestrel is entirely on the floor. And work on workaround / fix in CoreFX in parallel.

**Comment by benaadams** [comment]
The call back into sync `Stream.BeginWrite` [occurs here in SslStreamInternal](https://github.com/dotnet/corefx/blob/master/src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs#L453-L465)

**Comment by benaadams** [comment]
Sorry... @cesarbs  I think the fix would be for Kestrel to remove [its #ifdef s on its Streams here](https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameRequestStream.cs#L77-L115)?

**Comment by cesarbs** [comment]
I think that would require updating it to `netstandard1.7`, which we don't want to do.

**Comment by benaadams** [comment]
Extend ifdef for netstandard1.7 and netstandard2.0? Though it is a terrible code path...

**Comment by benaadams** [comment]
If SslStream implemented WriteAsync etc then it wouldn't call into the base Stream.WriteAsync which takes these paths.

**Comment by Priya91** [comment]
@cesarbs Can you please provide smaller repo outside of kestrel. It is not clear what the bug is.

**Comment by cesarbs** [comment]
I figured out what's happening. This is just a heads up, need a few minutes to write it up.

**Comment by cesarbs** [comment]
Sorry was on the way home and will have dinner now. Will be back soon.

**Comment by cesarbs** [comment]
In a nutshell, this is how Kestrel does HTTPS: on the read side, we have a task that repeatedly calls `SslStream.ReadAsync` and passes that up to whatever is consuming incoming data from the network. Writes are called directly either by the server itself or by the app running on it, and those eventually end up calling `SslStream.WriteAsync`.

Our `SslStream` wraps our own stream, [`LibuvStream`](https://github.com/aspnet/KestrelHttpServer/blob/c777a9efea7e8d633f8900cb2f55b0e56e569596/src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs). It essentially puts the `Stream` API on top of libuv's way of doing reads and writes.

This is what's happening in our read task:

1) `SslStream.ReadAsync()`
2) `Stream.ReadAsync()` (since `SslStream` does not override it)
3) `Stream.BeginEndReadAsync()`
4) Because `SslStream` (now) overrides `BeginRead`/`EndRead`, [this code path](https://github.com/dotnet/coreclr/blob/1cfbb8d79d65ea972eed8d73458fdf93697d031e/src/mscorlib/src/System/IO/Stream.cs#L430) is taken
5) The task started in `SslStream.BeginEndReadAsync()` calls `SslStream.BeginRead()`, which will ultimately call `LibuvStream.BeginRead()`
6) Because `LibuvStream` does not override `BeginRead()` (not on .NET Core), the [default implementation](https://github.com/dotnet/coreclr/blob/1cfbb8d79d65ea972eed8d73458fdf93697d031e/src/mscorlib/src/System/IO/Stream.cs#L301) is used
7) `Stream.BeginReadInternal()` (being called on `LibuvStream`) acquires a semaphore then proceeds to create a task that will do a [sync read](https://github.com/dotnet/coreclr/blob/1cfbb8d79d65ea972eed8d73458fdf93697d031e/src/mscorlib/src/System/IO/Stream.cs#L343) by calling `LibuvStream.Read()`. This will end up [here](https://github.com/dotnet/coreclr/blob/1cfbb8d79d65ea972eed8d73458fdf93697d031e/src/mscorlib/src/System/IO/Stream.cs#L343), where the thread will be blocked.

Now when the server or the app calls `WriteAsync`, the same chain of events will take place, but on write (`SslStream.WriteAsync` -> `Stream.WriteAsync` -> `Stream.BeginEndWriteAsync` -> etc etc). When it finally gets to `Stream.BeginWriteInternal()` being called on `LibuvStream`, it will block waiting to acquire the semaphore that is acquired on the read side of things. The read side will only release the semaphore when new data comes from the network on the same connection.

That's why it's hanging. My fix above (the pass through stream which only overrides `WriteAsync`) works because it causes the reading code to block on `PassThroughStream`'s semaphore, while writes will be using `LibuvStream`'s semaphore and therefore there will be no contention.

**Comment by cesarbs** [comment]
@benaadams Sorry, I misunderstood your suggestion earlier. Indeed, chaging the `#if` in [LibuvStream.cs](https://github.com/aspnet/KestrelHttpServer/blob/c777a9efea7e8d633f8900cb2f55b0e56e569596/src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs#L134) to `#if (NET451 || NETSTANDARD1_7)` is a possible fix. I'm not a fan of having both

```
    "netstandard1.3": {
      "dependencies": {
        "System.Threading.Thread": "4.4.0-*",
        "System.Threading.ThreadPool": "4.4.0-*"
      }
    }
```

and

```
    "netstandard1.7": {
      "dependencies": {
        "System.Threading.Thread": "4.4.0-*",
        "System.Threading.ThreadPool": "4.4.0-*"
      }
    }
```

in `project.json` though. And indeed it's not the ideal code path.

**Comment by benaadams** [comment]
Read and write using same semaphore :(

If it just needs WriteAsync overriding I think I have a fix for it; ReadAsync is messier...

**Comment by benaadams** [comment]
Raised issue in CoreClr https://github.com/dotnet/coreclr/issues/8752
Implemented `WriteAsync` on SslStream which should resolve this issue https://github.com/dotnet/corefx/pull/14771

**Comment by CIPop** [comment]
The PR above is only addressing the WriteAsync case. We need to address ReadAsync as well as NegotiateStream to consider this issue resolved.

We should also:
1. Add the test suggested by @cesarbs that [simulates ASP.Net's environment](https://github.com/dotnet/corefx/pull/14771#discussion_r94264181).
2. [Refactor WriteAsync ](https://github.com/dotnet/corefx/pull/14771#discussion_r94254974) to maximize code reuse as well as addressing some of the comments I've left in the original PR.
3. (Proposed offline by @Priya91) Figure out if the code in `Stream` that was previously performing an optimization is still a good idea now that we have APM APIs in the contracts.

Thanks @cesarbs and @benaadams for getting this blocking issue investigated and resolved in such a short time!

**Comment by benaadams** [comment]
ReadAsync is more complicated as it should use cached tasks; has to deal with reads smaller than the available size (if the read buffer is smaller); sync completes etc...

WriteAsync could probably use some optimization also; but the merged PR avoids Read and Write sharing the same semaphore and deadlocking https://github.com/dotnet/coreclr/issues/8752; don't know if that can be resolved in Stream.

**Comment by stephentoub** [comment]
> Because LibuvStream does not override BeginRead() (not on .NET Core)

Adding SslStream.WriteAsync/ReadAsync overrides that avoid all of the APM stuff and call WriteAsync/ReadAsync on the wrapped stream is a good thing to do, anyway, for performance reasons.  But I also think LibuvStream should override Begin/EndRead/Write.

**Comment by benaadams** [comment]
> I also think LibuvStream should override Begin/EndRead/Write.

It targets `netstandard1.3` though; Begin/End doesn't come in till `netstandard1.7`/ `netstandard2.0`

**Comment by stephentoub** [comment]
> It targets netstandard1.3 though; Begin/End doesn't come in till netstandard1.7/ netstandard2.0

I understand.  There should be a netstandard1.7/2.0 build.  It's a huge new version and deserves its own implementation for things like this.

**Comment by davidfowl** [comment]
That ends up being viral. Any library with a stream wrapper ends up having to cross compile...

**Comment by stephentoub** [comment]
> Any library with a stream wrapper ends up having to cross compile...

Yes.
